### ISSUE TITLE
feat: V6 增加亚像素 offset head，降低圆心定位误差

### ISSUE THREAD

**Issue by xiaoshuang112** [xiaoshuang112/circle-heat#1]
## Background

当前 `circle-heat` 的 V5 已完成类别矫正并重新训练。

当前已知基线：

- 有效样本：224
- occupancy：`occupied=120`、`empty=104`
- V5 最优 epoch：38
- 验证中心 mean：约 **2.79 px**
- 验证中心 P95：约 **5.07 px**
- occupancy accuracy：**100%**
- localizable：当前验证上 **100%**

现有中心头为：

```text
320×320 input
→ MobileNetV3-Small backbone
→ FPN
→ 80×80 heatmap
→ 局部峰值解码
```

当前 heatmap stride = 4，即 heatmap 1 个格对应输入图约 4 px。现场尺度约为：

```text
4 px ≈ 2 mm
1 px ≈ 0.5 mm
```

因此当前 mean/P95 对应的物理量已经不可忽略。现阶段没有证据说明 backbone 表达能力不足；空满分类已经稳定，剩余主要问题是**正确目标附近最后几像素的精定位**。

## Goal

在**不更换 backbone、不扩大输入尺寸、不改变 occupancy/localizable 任务定义**的前提下，为现有 80×80 heatmap 增加一个轻量亚像素 offset 定位头，使网络在相同数据和相同 split 上显著降低真实中心误差，同时保持现有空满分类能力和推理性能。

本轮目标不是追求更复杂网络，而是验证：

> 80×80 heatmap 负责粗定位，offset head 负责 4 px cell 内的细定位，是否能以极小计算代价换来有工程意义的像素精度提升。

## Non-goals

本轮不做：

- 不更换 MobileNetV3-Small；
- 不切换到 MobileNetV3-Large / ResNet / ConvNeXt / Transformer；
- 不把 heatmap 直接改成 160×160；
- 不修改 320×320 输入契约；
- 不增加 segmentation / radius / grid / LOO 等额外任务；
- 不修改 occupancy 标签和 `sigmoid >= 0.5` 现有逻辑；
- 不修改数据 split；
- 不引入新的复杂 loss 自动加权框架；
- 不直接改 `ai-det` 生产调用链；
- 不做 FP16 / INT8 优化。

## Current System

当前核心结构位于 `train.py`：

- `MobileNetV3-Small` pretrained backbone；
- FPN 最终得到约 `64×80×80` 特征；
- `heatmap_head = Conv2d(64, 1, 1)`；
- `localizable_head` 和 `occupancy_head` 基于最后一层全局池化；
- center heatmap 通过局部峰值邻域加权解码为连续坐标；
- 当前 heatmap stride 为 4。

现有 letterbox、标签坐标和逆变换必须保持 byte-level 行为一致，避免把 V5→V6 的收益混入坐标契约变化。

### 1. 保持 V5 主体不变

```text
320×320
   ↓
MobileNetV3-Small
   ↓
FPN 64×80×80
   ├─ heatmap head      → 1×80×80
   ├─ offset head NEW   → 2×80×80
   ├─ localizable head
   └─ occupancy head
```

### 2. 新增轻量 offset head

第一版保持最小实现：

```text
64×80×80
→ depthwise 3×3
→ pointwise 1×1
→ 2×80×80
```

两个通道分别为：

- `offset_x`
- `offset_y`

不增加额外 backbone，不建立第二阶段 ROI 网络。

### 3. Offset 监督定义

将真实中心映射到 heatmap 连续坐标：

```text
u = (x + 0.5) / 4 - 0.5
v = (y + 0.5) / 4 - 0.5
```

选定监督 cell：

```text
i = round(u)
j = round(v)
```

offset target：

```text
du = u - i
dv = v - j
```

理论范围约为 `[-0.5, 0.5]`。

Offset loss 只在 GT 对应 cell 上计算，不对整张 80×80 offset map 做密集监督。

第一版建议：

```text
L_offset = SmoothL1(pred_offset[j,i], gt_offset)
```

总 loss：

```text
L =
L_center
+ λ_offset * L_offset
+ 0.2 * L_localizable
+ current_occupancy_weight * L_occupancy
```

`λ_offset` 先取简单固定值（例如 1.0）并通过 V5/V6 A/B 验证，不做动态 loss weighting。

### 4. 推理解码

Heatmap 仍负责定位粗 cell。

推荐第一版严格使用：

```text
peak cell (i, j)
+ offset_head[:, j, i]
→ continuous heatmap coordinate
→ map back to 320 input
→ inverse letterbox
→ ROI coordinate
```

最终：

```text
u_final = i + du
v_final = j + dv

x = (u_final + 0.5) * 4 - 0.5
y = (v_final + 0.5) * 4 - 0.5
```

V6 不再依赖现有局部 soft-argmax 作为最终精定位结果；可保留其结果用于 A/B diagnostics。

不要同时引入多个新 decoder，以免无法判断提升来自哪里。

## Integration Contract

V6 训练和导出需要冻结以下契约：

### Input

- 单格位原始 bbox crop；
- 不外扩；
- 保持比例 letterbox 到 `320×320`；
- 与 V5 完全相同的 RGB / normalization。

### Outputs

建议 V6 ONNX 输出：

```text
heatmap_logits       [B, 80, 80]
offset_xy            [B, 2, 80, 80]
localizable_logit    [B]
occupancy_logit      [B]
```

不要在 ONNX 内嵌 letterbox / crop / 原图坐标变换。

### Coordinate semantics

- `heatmap_logits` 和 `offset_xy` 均基于 80×80 heatmap 坐标；
- offset 是相对 peak cell 的连续偏移；
- 最终原图坐标变换仍由现有部署侧完成。

### Experiment A — V5 baseline freeze

先固定并保存：

- V5 best.pt；
- 当前 train/val split；
- 当前标签版本；
- 当前 augmentation；
- 当前随机 seed；
- V5 验证逐样本预测。

输出：

- mean
- median
- P95
- max
- occupancy confusion matrix
- worst-N 样本列表

### Experiment B — V6 offset head

除新增 offset head / offset loss / decoder 外，其他全部保持一致。

使用相同：

- 224 样本；
- train/val split；
- augmentation；
- optimizer；
- LR schedule；
- early stop；
- random seed。

### Experiment C — Failure analysis

对 V6 最大误差样本按两类分开：

1. **目标结构选对，仅有小幅偏移**
   - offset head 应主要改善这一类。

2. **目标结构选错 / 热图 peak 已落在错误结构**
   - offset head 不应被期待修复；
   - 这类进入后续 hard-sample 数据迭代，不继续堆定位头。

## Acceptance Criteria

V6 必须满足：

### Accuracy

在与 V5 完全相同的 validation split 上：

- center mean 明显优于 V5 的约 2.79 px；
- center P95 明显优于 V5 的约 5.07 px；
- max error 不得出现明显退化；
- occupancy accuracy 不得低于 V5；
- localizable 行为不得出现非预期退化。

本 Issue 不预先写死“必须 P95 ≤ X px”，避免为了指标硬调；是否有工程价值以实际 A/B 为准。

优先期待的目标区间：

```text
mean ≈ 1.5–2.0 px
P95 ≈ 3 px 左右
```

这是实验目标，不是强制验收门槛。

### Performance

在当前 Orin / TensorRT FP32 环境：

- 记录 V5 与 V6 单格 GPU P50/P95；
- 当前 V5 参考值约 2.1 ms/slot；
- offset head 不应引入显著性能退化；
- 若 V6 明显增加 latency，必须说明其精度收益是否值得。

不要现在做 batch/FP16/INT8 优化来掩盖结构开销。

### Reproducibility

必须保留：

- V5 / V6 config；
- best epoch；
- best checkpoint；
- validation metrics；
- worst-N 可视化；
- 导出 ONNX 的输出 shape；
- 同一 validation set 上 V5/V6 对照表。

## Task DAG

```text
A. Freeze V5 baseline + per-sample predictions
                    ↓
B. Add offset head + target/loss
                    ↓
C. Add V6 decoder + unit/self-check
                    ↓
D. Train V6 on same split
                    ↓
E. V5 vs V6 A/B + worst-case analysis
                    ↓
F. Export ONNX + numerical parity check
                    ↓
G. Orin TensorRT FP32 benchmark
```

B/C 可以由同一 Agent 连续完成，不需要人为拆多个 PR；本 Issue 规模不值得为了并行增加协调成本。

## Definition of Done

本 Issue 完成必须有以下证据：

- [ ] V5 baseline 已冻结并可复现；
- [ ] offset head 已实现；
- [ ] offset target/loss 有最小单元测试或 self-check；
- [ ] decoder round-trip 坐标测试通过；
- [ ] V6 完成训练并按现有 early-stop 规则收敛；
- [ ] V5/V6 使用同一 validation split 完成 A/B；
- [ ] 输出 mean / median / P95 / max；
- [ ] 输出 occupancy confusion matrix；
- [ ] worst-N 样本完成可视化和失败分类；
- [ ] ONNX 导出并与 PyTorch 结果做数值一致性检查；
- [ ] 当前 Orin 上完成 FP32 TensorRT 单格 P50/P95 benchmark；
- [ ] 明确给出最终结论：V6 是否值得替换 V5。

## Stop / Escalation Rule

如果连续两轮只调整 offset loss / decoder 参数仍无法明显改善 P95：

**停止继续调 offset head。**

回到失败样本检查：

- 是否热图已经选错物理结构；
- 是否人工中心标签存在噪声；
- 是否 ROI / letterbox / 坐标还原存在系统偏差；
- 是否真正需要更多 hard samples。

不要进入“继续堆 offset 规则 / 阈值 / decoder”的补丁循环。

如果 V6 有明确收益，再单独开后续 Issue 集成到 `ai-det`；本 Issue 只负责把定位头实验和训练证据做干净。
